숫자야구 [Step 2] Rilla, Chirsty

christy-hs-lee opened the pull request:
@Gundy93 

안녕하세요 건디!  
두번째 PR 입니다! 이번에도 잘부탁드려요 😄👍🏻😄

## iOS-Number-Baseball [Step1] By Rilla, Christy

---

## 고민했던 부분💡

### readLine의 안전한 옵셔널 추출 방법
> readLine을 통해 받아온 문자열이 옵셔널이기 때문에 안전하게 추출하는 방법에 대해 고민을 많이 했었어요!
> 구글링을 통해 찾은 자료들 대부분이 강제 추출을 사용하여 해결했는데, 과제 제약사항에 위반되는 부분이라
> 다른 해결 방법을 찾아야 했습니다. 그러다 'compactMap'을 알게 되었습니다.
> compactMap은 nil을 자동으로 제거해줘 안전하게 옵셔널 바인딩을 할 수 있었습니다.
> 사용 방법은 map과 똑같은 클로져로 사용하면 됩니다! 😄

### enclosing catch is not exhaustive 에러
> 처음에 저희가 정의한 모든 에러 타입을 catch를 했음에도 불구하고 지속적으로 enclosing catch is not exhaustive 에러가
> 발생했습니다. 이를 해결하는 방법으로 추가적으로 catch문 이 하나 더 필요했습니다. 이는 정의하지 않은 에러가 발생할 수 있는
> 상황에 대비하여 작성해 줘야 했습니다. 그리고 건디가 추가적으로 공부해 보라고 했던 'localizedDescription'은
> 에러의 코멘트 역할을 한다는 것을 알게 되었습니다! 👍🏻

---

## 조언이 필요한 부분 ❓

### 오류 처리 코드의 적절한 위치
> checkAvailability 메서드를 빈배열 리턴하는 것에서 오류를 throw 하는것으로 변경한 후 현재  playGame 에서 오류처리를 하고 있습니다! 이 때문에 playGame 메서드의 코드 길이가 훨씬 늘어나게 됐는데 오류처리 부분은 따로 메서드를 빼서 정리를 하는게 나을까요 
> 아니면 현재처럼 붙여서 써도 될까요?

juun97 committed:
Refactor : Step2 코드로 전환

기존 메인 코드 Step2 코드로 전환과정

christy-hs-lee committed:
feat : 에러 타입 정의

에러 타입을 정의함

juun97 committed:
refacor : checkAvailability() 메서드 수정

기존 빈배열 리턴에서 오류를 던지는 로직으로 수정

christy-hs-lee committed:
refactor : 에러 메세지 작성

각 에러에 대한 출력문 작성

christy-hs-lee committed:
docs : README step2 flow chart 수정

Gundy93 commented on README.md:
```diff
@@ -8,5 +8,5 @@
 
 ### Step2
-![image](https://user-images.githubusercontent.com/61936306/208640886-c59a79ff-2860-4a3e-a1f5-ff84c7f98940.png)
+![image](https://user-images.githubusercontent.com/61936306/209061174-dacce864-f817-4a5c-82e4-871ab23a9531.png)
```
|표준 배경|다크 모드|회색 배경|
|:-:|:-:|:-:|
|![](https://i.imgur.com/u5NAcKB.png)|![](https://i.imgur.com/CfmghSD.png)|![](https://i.imgur.com/6hvbWP8.png)|

배경이 투명한 상태라서 선이 어떻게 연결되는지 다크모드에서 확인하기 어렵네요🤔

|초기값|초깃값|
|:-:|:-:|
|![](https://i.imgur.com/dGkXPyA.png)|![](https://i.imgur.com/uQOygfW.png)|

단어의 통일이 필요할 것 같습니다😃

Gundy93 commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -23,28 +31,56 @@ func makeRandomNumbers() -> [Int] {
+            print("입력 형식이 잘못 되었습니다! 띄어쓰기를 준수해 주세요.")
+        } catch {
+            print(error.localizedDescription)
+        }
```
다양한 에러에 대해 핸들링을 해주신 부분 아주 잘하셨습니다! 👍

이번 프로젝트에서는 들여쓰기 2번의 제한이 있기 때문에 지금 로직 그대로는 활용할 수 없지만
`switch` 구문을 통해서도 오류처리를 할 수 있습니다

[Control Flow](https://docs.swift.org/swift-book/LanguageGuide/ControlFlow.html) 문서를 통해 다양한 흐름제어를 공부해보면 어떨까요?
`switch`는 `Conditional Statements` 파트에 소개되어 있습니다 😃

Gundy93 commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -9,6 +9,14 @@ import Foundation
 let numbers = Array(1...9)
```
해당 변수는 전역 변수일 필요가 없어 보이네요!
필요한 부분에서 사용할 수 있도록 지역 변수로 옮기는 것이 적절해보이는데,
여러분 생각은 어떠신가요? 🤔

juun97 replied:
numbers 배열이 makeRandomNumbers() 에서만 사용이 되어서 이 함수 안에서 지역변수로 쓰는걸로 변경했습니다!

Gundy93 commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -23,28 +31,56 @@ func makeRandomNumbers() -> [Int] {
+            print(error.localizedDescription)
+        }
+        
+        guard inputNumbers.isEmpty == false else {
+            continue
+        }
```
`inputNumbers`가 빈 배열인지를 확인해서 비어있다면 `continue`를 하게끔 하셨군요!
흐름제어에 대해 신경 쓰신 부분이 보입니다 👍
하지만 위의 `do-catch`문을 통과했을 때 `inputNumbers`가 비어있는 경우만이 잘못된 경우일까요?
이 부분에 대해서는 고민을 조금 해봐야할 것 같습니다 🤔

juun97 replied:
저희가 생각했을땐 checkAvailability()에서 넘어오는 형태가 Error 이거나 정상적인 값이 오거나 둘중 하나 일 수 밖에 없어서 값이 있는지 없는지 두개의 판별하는 방법밖에 떠오르지가 않아요 건디! 혹시 어떤 부분이 더 있을까요???

Gundy93 replied:
기존 로직대로는 `inputNumbers`가 `while`문 바깥에 위치했기 때문에 `do-catch`문을 통과했을 때 `inputNumbers`가 비어있는 경우만이 잘못된 경우는 아니었었죠!
하지만 `inputNumbers`를 `while`문 내부로 위치를 옮기면서 `inputNumbers`가 비어있는 경우만이 잘못된 경우가 됐기 때문에 같이 해결됐습니다 😃로직을 잘 수정하셨습니다! 👍

Gundy93 commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -23,28 +31,56 @@ func makeRandomNumbers() -> [Int] {
 func playGame() {
     var leftChances = 9
+    var inputNumbers : [Int] = []
     var result : [Int] = []
```
`inputNumbers`와 `result`는 `while`문 내부에서만 사용되는 값으로 보입니다
선언하는 위치가 `while`문 내부에 있는 것과 외부에 있는 것 중 어느 쪽이 적절한지 고민해봅시다 😃
또, 타입 어노테이션의 컨벤션이 약간 어색해 보이네요

Gundy93 commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -23,28 +31,56 @@ func makeRandomNumbers() -> [Int] {
+        
+        result = checkStrikeBall(numbers: inputNumbers)
+        
         let strike = result[0]
         let ball = result[1]
```
지난 step에서는 넘어갔던 부분인데요! 인덱스로 값에 접근하는 것은 위험할 수 있습니다
그 이유에 대해서 알고 계실까요? 한번정도 고민해보는게 좋겠습니다 😃

juun97 replied:
고민해본 결과 인덱스로 접근했을 때 유효하지 않은 인덱스로 접근하게되면 런타임오류가 발생할 위험이 있기 때문인거 같습니다!

Gundy93 commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -23,28 +31,56 @@ func makeRandomNumbers() -> [Int] {
-
+        
         if strike == 3 {
-            print("정답 입니다!")
+            print("사용자 승리!")
             return
         }
```
`if`문을 통해 특정 상황에서만 함수가 종료되도록 흐름제어에 신경쓰신 부분이 아주 좋습니다! 👍
이런 상황에서 `guard`문 보다 `if`문이 적절하다고 생각하신 이유에 대해서 알 수 있을까요?

juun97 replied:
아 이부분은 빠른종료가 목적이라 guard 문이 더 적절한거 같습니다! 저희가 놓쳤던 부분이네요 감사해요 건디 !!👍🏻👍🏻👍🏻

juun97 replied:
이부분은 빠른종료를 위해 guard 문이 더 적절해보이네요 건디! 저희가 놓친 부분입니다 감사해요 건디 👍🏻👍🏻👍🏻👍🏻

Gundy93 replied:
아! 저는 `if`문이 더 좋을 것 같습니다.
정말로 잘 사용하셨는데, `if`를 쓴 이유에 대해서 어떤 생각을 하고 사용하셨는지 궁금해서 여쭤본 것이었습니다!

christy-hs-lee replied:
저희가 if문을 사용한 이유는 특정한 하나의 상황에 적용하는 조건문을 만들어 가독성을 고려한 선택이었습니다!
guard문은 특정한 상황이 아닐 때 로직을 처리하고 싶을 때 사용하면 적절할 것 같습니다!

Gundy93 commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -23,28 +31,56 @@ func makeRandomNumbers() -> [Int] {
-
-        print("임의의 수 : \(randomNumbers[0]), \(randomNumbers[1]), \(randomNumbers[2])")
+        
         print("\(strike) 스트라이크, \(ball) 볼")
         print("남은 기회 : \(leftChances)")
+        print(" ")
```
스트라이크와 볼, 남은 기회에 대해 출력한 뒤에 `print(" ")`를 통해 한 줄 띄우고자 하신 부분 아주 좋았고 인상깊었습니다!

하지만 꼭 `" "`처럼 공백을 추가하지 않더라도 `print("")` 처럼 빈 값을 출력하거나 `\n`을 사용해보는 것도 좋겠네요

`print`함수에 대해서도 공부해볼까요?

christy-hs-lee replied:
print문 에서 사용하는 terminator는 default 값이 \n 이므로 \n\n로 작성해서 개행 처리 과정을 한줄로 축약시켰습니다!

Gundy93 commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -58,4 +94,59 @@ func checkResult(numbers: [Int]) -> [Int] {
-playGame()
+func checkAvailability(inputData: String?) throws -> [Int] {
+    
+    guard let inputData = inputData?.split(separator: " ") else {
+        throw inputError.wordSpacing
+    }
```
`wordSpacing` 에러가 발생하는 경우가 있을까요? 저는 매개변수의 타입이 `String?`이 아니라 `String`인 것이 더 적절할 것 같습니다 처음에 사용자 입력을 받을 때 옵셔널 처리를 해서 매개변수로 전달하는 것이 어떨까요?
여러분의 생각이 궁금합니다 😃

juun97 replied:
띄어쓰기를 하지 않은경우 숫자의 범위를 초과한것으로 뜨기 때문에 wordSpacing 오류를 삭제하고 처음 입력받을때 옵셔널 처리를 해서 String?이 아닌 String으로 받게 수정했습니다!

Gundy93 commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -23,28 +31,56 @@ func makeRandomNumbers() -> [Int] {
+        print("중복 숫자는 허용하지 않습니다.")
+        print("입력 : ", terminator: " ")
+        
+        let inputData = readLine()
```
`inputData`라는 네이밍은 변수의 타입이 `Data` 타입일 것 같은 느낌을 준다고 생각합니다
혹시 더 좋은 네이밍은 없을까요? 🤔

christy-hs-lee replied:
사용자의 입력값이라는 의미의 userInput 으로 변경 했습니다!

Gundy93 replied:
좋은 네이밍인것 같아요! 👍

Gundy93 commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -58,4 +94,59 @@ func checkResult(numbers: [Int]) -> [Int] {
+        throw inputError.numberOfInput
+    }
+    
+    let inputNumbers = inputData.compactMap { str in Int(str) }
```
[Naming](https://www.swift.org/documentation/api-design-guidelines/#naming)의 Use Terminology Well를 참고해서 클로져 부분을 수정할 수 있을까요?

juun97 replied:
str 을 보다 직관적이게 number 로 수정했습니다!

Gundy93 commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -58,4 +94,59 @@ func checkResult(numbers: [Int]) -> [Int] {
+        print("1. 게임시작")
+        print("2. 게임종료")
+        print("원하는 기능을 선택해주세요 : ", terminator: " ")
+        guard let inputData = readLine() else {
+            print("입력이 잘못되었습니다")
+            continue
+        }
```
`guard`문을 통해 `Optional`을 안전하게 처리 하셨네요 👍
하지만 `else` 블록 내부는 실행되지 않을 것 같아서 조금 손 보면 코드의 길이를 줄일 수 있겠네요 😃

juun97 replied:
else 블록 내는 실행되지 않아 print 문을 삭제했습니다!

Gundy93 commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -58,4 +94,59 @@ func checkResult(numbers: [Int]) -> [Int] {
+            continue
+        }
+        
+        switch inputData {
+        case "1" :
+            playGame()
+        case "2" :
+            print("게임을 종료합니다")
+            return
+        default :
+            print("입력이 잘못되었습니다")
+            continue
+        }
```
`switch`문을 통해 원하는 입력값이 아닌 경우 모두 `default`를 향하게 했군요!
아주 보기 좋습니다 LGTM!! 👍

Gundy93 reviewed (commented):
여러분 이번 스텝도 훌륭하게 작성하셨군요! 아주 만족스럽습니다 👍
제게 보내시는 마지막 PR인 만큼 더 꼼꼼히 보려고 노력했습니다
앞으로도 열심히 하셨으면 좋겠습니다 😃

## 고민했던 부분💡
### readLine의 안전한 옵셔널 추출 방법
Swift는 안전성을 추구하기 때문에 안전한 옵셔널 추출 방법에 대해 고민하신 부분 아주 좋습니다!

### enclosing catch is not exhaustive 에러
`localizedDescription`에 대해 공부해보셨군요! 잘하셨습니다 👍
오류에 대해서 공부하시는 김에 `LocalizedError`에 대해서도 알아보시면 좋을 것 같네요

---

## 조언이 필요한 부분 ❓
### 오류 처리 코드의 적절한 위치
기능 분리에 대해 고민을 하셨군요! 바람직한 고민입니다
함수는 되도록이면 하나의 기능을 하는 것이 적절하기 때문에 함수에서 요구하는 기능 외적인 부분은 다른 함수로 분리하는 것이 적절하다고 대답할 수 있겠네요

추가로 여러분이 `메서드`라고 표현해주셨는데요, `메서드`와 `함수`의 차이에 대해서 알고계신지 궁금하네요 🤔
한번 공부해보시는 것도 좋을 것 같습니다 😃

---

### 사용자 입력 받을 때
![](https://i.imgur.com/AJks8xR.png)

위의 이미지와 같이 해당 부분은 콜론 뒤로 공백이 두번 들어가네요! 이 공백을 한칸으로 줄이면 더 좋을 것 같습니다
이번 스텝의 주요 기능과는 무관한 부분이라 크게 신경쓰지 않으셔도 될 것 같네요 😃

### import Foundation
이번 코드는 step1과 달리 `import Foundation`이 필요해 보이네요!
과연 무엇때문에 이 `import Foundation`이 필요한지 혹시 알고 계실까요?
생각해보시고 추가적인 코멘트로 답변해주시면 좋을 것 같습니다 😃

### 오류 처리
![](https://i.imgur.com/pErM96H.png)

이렇게 중복값 및 다양한 입력에 대해 에러처리를 시도한 부분 훌륭합니다 👍

하지만 중간에 예상치 못한 에러를 발견할 수 있었는데요

![](https://i.imgur.com/YRUQS30.png)

이러한 오류를 발견하였는데 어떻게 로직을 수정하면 좋을지 생각해보는 것이 좋겠습니다

**오류 발생 조건**
- 카운트가 차감되기 전에는 잘 동작하나, 카운트가 차감되는 순간부터는 잘못된 형식의 값일 경우 앞선 입력값을 통해 체크가 이루어지면서 에러문구와 기회 차감이 동시에 발생합니다

해결하는 방법은 여러가지가 있을 것 같은데, 천천히 생각해보고 잘 모르겠는 부분에 대해서는 말씀해주세요 🙌
혹시 힌트가 필요하다면 아래의 코멘트 중에 있습니다 😃 화이팅 💪

juun97 committed:
fix : playGame() 수정

카운트가 차감된 후 오류 출력이 되어도 스트라이크와 볼이 출력되고 남은기회가 차감되는 오류를 수정했습니다. inutNumbers 배열을 while 문 안에서 선언하여 while 문이 돌때마다 refresh 되도록 수정했습니다.

juun97 committed:
Merge branch 'main' of https://github.com/christy-hs-lee/ios-number-baseball

christy-hs-lee committed:
refactor : numbers 전역변수 수정

전역변수를 지역변수로 설정

juun97 committed:
refactor : playGame() 함수 수정

변수의 선언위치 컨벤션 수정, readLine의 옵셔널 처리, 개행처리 그리고 if 문을 guard문으로 수정했습니다

juun97 committed:
Revert "refactor : playGame() 함수 수정"

This reverts commit 4c2471c6d913081744397be1e028323b30f0206b.

juun97 committed:
chore : empty commit

juun97 committed:
refactor : playGame 함수 수정

어노테이션 컨벤션 수정, if 문을 guard문으로 변환, 변수명 수정, print 문 개행처리 수정

christy-hs-lee committed:
refactor : checkAvailability 함수 수정 및 에러 타입 수정

checkAvailability 함수  정규식 수정 및 에러 타입 정리
wordSpacing 에러 타입 삭제
변수명 수정

juun97 committed:
refactor : checkAvailability() 함수 수정

str 을 직관적으로 볼 수 있게 number 로 수정했습니다

christy-hs-lee committed:
fefactor : menuPrint 함수 수정

변수명 수정 및 guard문 속 불필요한 출력문 삭제

christy-hs-lee committed:
docs : README step2 flow chart 수정

christy-hs-lee committed:
refactor : playGame 함수 수정

사용자 승리를 출력하는 부분 수정

christy-hs-lee committed:
Merge branch 'main' of https://github.com/christy-hs-lee/ios-number-baseball into main

juun97 committed:
refactor : checkAvailability() 수정

숫자가 아닌값을 입력받았을때 오류 처리하는 부분을 수정했습니다!

juun97 committed:
refactor : checkAvailability() 수정

내부 로직 수정

Gundy93 commented:
여러분 리팩토링까지 훌륭히 마무리해주셨군요~ 이번주 열심히 하시는 모습 보기 좋았습니다 앞으로도 화이팅하는 캠프 생활이 되기를 바랍니다! 앞으로도 편하게 연락주시기를 기대할게요 😃
